File: box-and-box/ports/go/value.go
package main

// value.go — Invariant Arithmetic (port of value.mjs).
// A Value is a PRODUCT OF MONOIDS across families. combine merges; chain composes
// along PULSE phases (partial — refuses a backward step); promote/reconcile/deliberate
// are endomorphisms; consume is the boolean gate. Laws L1–L14.

import (
	"fmt"
	"math"
)

var Phases = []string{"retrieve", "route", "act", "learn", "consolidate"}

func phaseIdx(p *string) int {
	if p == nil {
		return -1
	}
	for i, x := range Phases {
		if x == *p {
			return i
		}
	}
	return -1
}

// Value is the carrier of the invariant product monoid.
type Value struct {
	N           float64  // ℝ under +
	Kappa       bool     // Bool under ∨ (OR) — cyclicity
	Beta        float64  // [0,1] under min — persistence / confidence
	Sigma       []string // Set<Tag> under ∪ — derived conflicts
	Pi          *string  // Phase|nil, first-non-null (NOT commutative)
	Iota        *string  // IdemKey, first-non-null
	Psi         *string  // Cadence, first-non-null
	Authority   []string // List<Cap> under concat (free monoid)
	DenyDefault bool     // Bool under ∧ (AND)
	Audit       []string // List<Event> under concat (free monoid)
	Err         string   // non-empty marks a ⟂ error value (from chain)
}

// V0 — identity element of the whole product monoid.
func V0() Value {
	return Value{
		N:           0,
		Kappa:       false,
		Beta:        1,
		Sigma:       []string{},
		Pi:          nil,
		Iota:        nil,
		Psi:         nil,
		Authority:   []string{},
		DenyDefault: true,
		Audit:       []string{},
	}
}

func strPtr(s string) *string { return &s }

func uniq(arr []string) []string {
	seen := map[string]bool{}
	out := []string{}
	for _, x := range arr {
		if !seen[x] {
			seen[x] = true
			out = append(out, x)
		}
	}
	return out
}

func firstNonNull(a, b *string) *string {
	if a != nil {
		return a
	}
	return b
}

func concatStr(a, b []string) []string {
	out := make([]string, 0, len(a)+len(b))
	out = append(out, a...)
	out = append(out, b...)
	return out
}

// combine : Value × Value → Value. Componentwise monoid op.
func combine(a, b Value) Value {
	return Value{
		N:           a.N + b.N,
		Kappa:       a.Kappa || b.Kappa,
		Beta:        math.Min(a.Beta, b.Beta),
		Sigma:       uniq(concatStr(a.Sigma, b.Sigma)),
		Pi:          firstNonNull(a.Pi, b.Pi),
		Iota:        firstNonNull(a.Iota, b.Iota),
		Psi:         firstNonNull(a.Psi, b.Psi),
		Authority:   concatStr(a.Authority, b.Authority),
		DenyDefault: a.DenyDefault && b.DenyDefault,
		Audit:       concatStr(a.Audit, b.Audit),
	}
}

// chain : Value × Value → Value (PARTIAL). Defined only when phase(a) ≤ phase(b).
func chain(a, b Value) Value {
	if a.Pi != nil && b.Pi != nil && phaseIdx(a.Pi) > phaseIdx(b.Pi) {
		return Value{Err: fmt.Sprintf("π-violation: cannot chain '%s' after '%s'", *b.Pi, *a.Pi)}
	}
	r := combine(a, b)
	r.Pi = firstNonNull(b.Pi, a.Pi) // exit phase
	return r
}

// Evidence for promote.
type Evidence struct {
	Beta    float64
	HasBeta bool
}

// promote : β-monotone endomorphism: promote(v).β ≥ v.β.
func promote(v Value, ev Evidence) Value {
	r := cloneValue(v)
	eb := 0.0
	if ev.HasBeta {
		eb = ev.Beta
	}
	r.Beta = math.Max(v.Beta, eb)
	return r
}

func cloneValue(v Value) Value {
	r := v
	r.Sigma = append([]string{}, v.Sigma...)
	r.Authority = append([]string{}, v.Authority...)
	r.Audit = append([]string{}, v.Audit...)
	return r
}

// reconcile : σ-antitone, idempotent endomorphism: removes resolved conflict tags.
func reconcile(v Value, tags []string) Value {
	drop := map[string]bool{}
	for _, t := range tags {
		drop[t] = true
	}
	r := cloneValue(v)
	out := []string{}
	for _, t := range v.Sigma {
		if !drop[t] {
			out = append(out, t)
		}
	}
	r.Sigma = out
	return r
}

// deliberate : κ-antitone, idempotent endomorphism: forces κ = false.
func deliberate(v Value) Value {
	r := cloneValue(v)
	r.Kappa = false
	return r
}

// Requirements for consume.
type Requirements struct {
	BetaMin     *float64
	SigmaEmpty  bool
	Acyclic     bool
	Phase       *string
	ForwardFrom *string
	DenyDefault string // "must_allow" or ""
	Authorized  bool
}

// Failure is one gate failure.
type Failure struct {
	Family string
	Why    string
}

// Verdict is the result of consume.
type Verdict struct {
	Ok       bool
	Failures []Failure
	Value    Value
}

// consume : the correctness gate (a predicate, not an operation on Value).
func consume(v Value, req Requirements) Verdict {
	failures := []Failure{}
	if req.BetaMin != nil && v.Beta < *req.BetaMin {
		failures = append(failures, Failure{"β", fmt.Sprintf("β=%v < β_min=%v", round3(v.Beta), *req.BetaMin)})
	}
	if req.SigmaEmpty && len(v.Sigma) > 0 {
		failures = append(failures, Failure{"σ", "unresolved conflicts"})
	}
	if req.Acyclic && v.Kappa {
		failures = append(failures, Failure{"κ", "cyclic — self-reference detected"})
	}
	if req.Phase != nil && (v.Pi == nil || *v.Pi != *req.Phase) {
		piStr := "nil"
		if v.Pi != nil {
			piStr = *v.Pi
		}
		failures = append(failures, Failure{"π", fmt.Sprintf("phase %s ≠ required %s", piStr, *req.Phase)})
	}
	if req.ForwardFrom != nil && v.Pi != nil && phaseIdx(v.Pi) < phaseIdx(req.ForwardFrom) {
		failures = append(failures, Failure{"π", fmt.Sprintf("phase %s precedes %s", *v.Pi, *req.ForwardFrom)})
	}
	if req.DenyDefault == "must_allow" && v.DenyDefault && !req.Authorized {
		failures = append(failures, Failure{"governance", "deny_default with empty authority_path"})
	}
	return Verdict{Ok: len(failures) == 0, Failures: failures, Value: v}
}

func round3(x float64) float64 { return math.Round(x*1000) / 1000 }

File: box-and-box/ports/python/value.py
"""value.py - Invariant Arithmetic, faithful Python port (v0.2)

A Value is a PRODUCT OF MONOIDS across families. ``combine`` merges; ``chain``
composes along PULSE phases (partial - refuses a backward step); ``promote`` /
``reconcile`` / ``deliberate`` are endomorphisms; ``consume`` is the boolean gate.
These satisfy laws L1-L14 (see laws.py). Semantics mirror value.mjs exactly.

A Value is represented as a plain ``dict`` with the keys below. ``chain`` may
return ``{"error": ...}`` when a backward phase step is refused.
"""

PHASES = ["retrieve", "route", "act", "learn", "consolidate"]


def phase_idx(p):
    return PHASES.index(p) if p in PHASES else -1


def _uniq(arr):
    # preserve first-seen order, dedupe (set semantics for sigma)
    out = []
    seen = set()
    for x in arr:
        if x not in seen:
            seen.add(x)
            out.append(x)
    return out


def _first_non_null(a, b):
    return a if a is not None else b


def _clone(v):
    r = dict(v)
    r["sigma"] = list(v["sigma"])
    r["authority"] = list(v["authority"])
    r["audit"] = list(v["audit"])
    return r


def V0():
    """Identity element of the whole product monoid."""
    return {
        "n": 0,             # R under +
        "kappa": False,     # Bool under OR - cyclicity
        "beta": 1,          # [0,1] under min - persistence / confidence
        "sigma": [],        # Set<Tag> under union - derived conflicts
        "pi": None,         # Phase|None, first-non-null (NOT commutative)
        "iota": None,       # IdemKey, first-non-null
        "psi": None,        # Cadence, first-non-null
        "authority": [],    # List<Cap> under concat (free monoid)
        "denyDefault": True,  # Bool under AND
        "audit": [],        # List<Event> under concat (free monoid)
    }


def V(p=None):
    """Build a Value with sensible defaults."""
    p = p or {}
    v = V0()
    v.update(p)
    v["sigma"] = list(p.get("sigma", []))
    v["authority"] = list(p.get("authority", []))
    v["audit"] = list(p.get("audit", []))
    return v


def combine(a, b):
    """combine : Value x Value -> Value.

    Componentwise monoid op. NOT globally commutative (temporal & governance
    are first-non-null / concat), but associative with identity V0 => a monoid.
    """
    return {
        "n": a["n"] + b["n"],
        "kappa": a["kappa"] or b["kappa"],
        "beta": min(a["beta"], b["beta"]),
        "sigma": _uniq(list(a["sigma"]) + list(b["sigma"])),
        "pi": _first_non_null(a["pi"], b["pi"]),
        "iota": _first_non_null(a["iota"], b["iota"]),
        "psi": _first_non_null(a["psi"], b["psi"]),
        "authority": list(a["authority"]) + list(b["authority"]),
        "denyDefault": a["denyDefault"] and b["denyDefault"],
        "audit": list(a["audit"]) + list(b["audit"]),  # pure concat (free monoid)
    }


def chain(a, b):
    """chain : Value x Value -> Value (PARTIAL).

    Defined only when phase(a) <= phase(b) in PULSE order; a backward step is
    REFUSED (returns {"error": ...}). Composition moves the value to the exit
    (later) phase.
    """
    if a["pi"] is not None and b["pi"] is not None and phase_idx(a["pi"]) > phase_idx(b["pi"]):
        return {"error": f"pi-violation: cannot chain '{b['pi']}' after '{a['pi']}'"}
    r = combine(a, b)
    r["pi"] = _first_non_null(b["pi"], a["pi"])  # exit phase
    return r


def promote(v, evidence=None):
    """promote : Value x Evidence -> Value.

    beta-monotone endomorphism: promote(v).beta >= v.beta, always.
    """
    evidence = evidence or {}
    r = _clone(v)
    ev_beta = evidence.get("beta")
    ev_beta = ev_beta if ev_beta is not None else 0
    r["beta"] = max(v["beta"], ev_beta)
    return r


def reconcile(v, tags=None):
    """reconcile : Value x Set<Tag> -> Value.

    sigma-antitone, idempotent endomorphism: removes resolved conflict tags.
    """
    tags = tags or []
    drop = set(tags)
    r = _clone(v)
    r["sigma"] = [t for t in v["sigma"] if t not in drop]
    return r


def deliberate(v):
    """deliberate : Value -> Value.

    kappa-antitone, idempotent endomorphism: forces kappa = False.
    """
    r = _clone(v)
    r["kappa"] = False
    return r


def consume(v, req=None):
    """consume : Value x Requirements -> {ok, failures, value}.

    The correctness gate (a predicate, not an operation on Value).
    """
    req = req or {}
    failures = []
    if req.get("beta_min") is not None and v["beta"] < req["beta_min"]:
        failures.append({"family": "beta", "why": f"beta={_round(v['beta'])} < beta_min={req['beta_min']}"})
    if req.get("sigma_empty") and len(v["sigma"]) > 0:
        failures.append({"family": "sigma", "why": f"unresolved conflicts {{{', '.join(map(str, v['sigma']))}}}"})
    if req.get("acyclic") and v["kappa"]:
        failures.append({"family": "kappa", "why": "cyclic - self-reference detected"})
    if req.get("phase") and v["pi"] != req["phase"]:
        failures.append({"family": "pi", "why": f"phase {v['pi']} != required {req['phase']}"})
    if req.get("forward_from") and v["pi"] is not None and phase_idx(v["pi"]) < phase_idx(req["forward_from"]):
        failures.append({"family": "pi", "why": f"phase {v['pi']} precedes {req['forward_from']}"})
    if req.get("deny_default") == "must_allow" and v["denyDefault"] is True and req.get("authorized") is not True:
        failures.append({"family": "governance", "why": "deny_default with empty authority_path"})
    return {"ok": len(failures) == 0, "failures": failures, "value": v}


def _round(x):
    return round(x * 1000) / 1000


def digest(v):
    if v.get("error"):
        return f"_|_ {v['error']}"
    return "  ".join([
        f"n={_round(v['n'])}",
        f"kappa={v['kappa']}",
        f"beta={_round(v['beta'])}",
        f"sigma={{{','.join(map(str, v['sigma']))}}}",
        f"pi={v['pi'] if v['pi'] is not None else '.'}",
        f"auth=[{'.'.join(map(str, v['authority']))}]",
        f"deny={v['denyDefault']}",
    ])
